[Search] Decouple SearchFieldCategory from DocumentFieldCategory.cs

c3dr0x opened the pull request:
Fix #18

**[BC]** Split `SearchFieldCategory` in two : `DocumentFieldCategory` & `SearchFieldCategory`

### DocumentFieldCategory

Is used to detect which are the fields to use as `Id`, `Search` and `Security`

### SearchFieldCategory

Is used to know how to store/index & then query fields

New types are :

-  Result
-  TextSearch
-  Sort
-  Term
-  ListTerm
-  None

**Note** : `Filter`, `Facet` and `ListFacet` no longer exists. 

## [BC] 

Fields still need to have a `SearchFieldAttribute` so you need to update them as follow :
- **Search** : replace with `[DocumentField(DocumentFieldCategory.Search)]` and `[SearchField(SearchFieldCategory.TextSearch)]`
- **Security** : replace with `[DocumentField(DocumentFieldCategory.Security)]` and `[SearchField(SearchFieldCategory.XXX)]`. it depends, you are now free to use the one you need
- **Id** : replace with `[DocumentField(DocumentFieldCategory.Id)]` and `[SearchField(SearchFieldCategory.None)]`. haven't come with a better one
- **Facet** : replace with `[SearchField(DocumentFieldCategory.Term)]`
- **Filter** : replace with `[SearchField(DocumentFieldCategory.Term)]`
- **ListFacet** : replace with `[SearchField(DocumentFieldCategory.ListTerm)]`
- **Result** : stays the same
- **Sort** : stays the same

## Benefit

`Filter` can now be done with : `Term`, `ListTerm` and `TextSearch`. Before you could hack by using `ListFacet` but they was no way to use a `TextSearch` like.

JabX commented:
`SearchText` and `ListTerm` are the same and should probably be combined and named `Terms`, because it is already the denomination that ES uses.

I also wonder what use does `None` has, both in terms of "why do I need to type this instead of not giving an attribute ?" and "what can I do in ES with a field that I marked as `None` ?". If you don't specify anything it's going to use the default values for indexing and storing, (which I think is the same as `Result`?), so it's definitely not "not indexed", which is the meaning I can deduce from `None`.

durandx commented:
Some remarks: 
 1. Don't forget to impact Kinetix.Search.Test : The tests should compile and pass.
 2. Why not simply add a TextSearch in SearchFieldCategory instead of splitting this class ?
 3. Is None really usefull ? if yes, in which case ?
 4. I think "Facet" and "ListFacet" should be kept as their names are meaningfull and used in others API (ex: QueryInput, QueryOutput, ...)

c3dr0x commented:
1. I missed about Kinetix.Search.Test, will do.

2. I want to split it in two because I think they should have been from the beginning. 
`Id`, `Search `and `Security` are about logic about the document, not about how the column should be indexed.

For example `Security` on some documents can be one Id or a list of Id, so the way it is indexed changes but not the logic that this is the `Security` field to match against. I had this case on previous projects and I worked my way around by adding a `ListSecurity`.

So the issue by keeping them merged for me is having duplication of behavior just to keep one list. 

3. `None` is not very useful indeed, I added it to keep the fact that a `SearchFieldAttribute` is mandatory on all fields and it matched the previous case for `Id`. But you and @JabX seems to be disturbed by it, so it is probably a bad idea.

4. I don't agree that they have real meaning, all projects today when they want to do a `ListFilter` uses `ListFacet` on Filter fields as a hack. Kinetix only uses them to create the document in ES but never when querying. Eg: force Fitlers to be Filter and Facets to be Facet.

So keeping `ListFacet` and `Facet` would force to add `Filter`, `ListFilter`, `TextFilter` to be coherent.
If I renamed it, it's to be closer to the ES terminology.

JabX commented:
"Facet" and "ListFacet" have no meaning at all for the API, it's purely internal. "ListFacet" exists to fulfill the need of having a document that matches several facet values (instead of only one), and facets in general are described in a "FacetQueryFactory" which is totally separate from whatever attributes you put on the document fields.

JabX commented:
Since this PR has been closed for some reason (I guess inactivity), I have been thinking about this again.

It turns out (because it wasn't clear for me before) that storing fields in ES (as in `{store: true}`) is virtually useless because ES stores the entire document in the `_source` property and is what's used to retrieve the documents when they are queried. I think the uses cases for storing a property are extremely limited and very far from the scope we're covering here, so we shouldn't even bother with it.

That being said, we can pretty much keep what was proposed here but we should rename `SearchFieldCategory` into `SearchFieldIndexing` (because it only controls how the property is indexed, if it even is), rename `Result` to `None` and have it set as `{index: false, store: false}`.

The possible values for `SearchFieldIndexing` would be:
- `None` (default)
- `Term` (== keyword)
- `Terms` (== tokenized text with fielddata)
- `Sort` (== lowercased keyword)
- `FullText` (== tokenized text)

and the possible values for `SearchFieldCategory` (formally `DocumentFieldCategory` in this PR but now the original name is free again) would be:
- `None` (default)
- `Id`
- `SearchText`
- `Security`

Both `None` being default values mean that we can simply omit to specify them in the documents if the property isn't indexed or isn't special, and that has two upsides:
- There's no need to specify `[SearchField(SearchFieldCategory.Result)]` everywhere
- It becomes obvious that it's totally okay to use the same field for filtering/facetting and as a result.

To finish, I would suggest that we use the same `SearchField` annotation to specify both attributes, we just need like two constructors to handle both attribute order.
